rest/src/main/java/io/syndesis/rest/v1/state/ClientSideState.java

```diff
@@ -0,0 +1,291 @@
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.syndesis.rest.v1.state;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.GeneralSecurityException;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Base64.Decoder;
+import java.util.Base64.Encoder;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import javax.crypto.Cipher;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.ws.rs.core.Cookie;
+import javax.ws.rs.core.NewCookie;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.ObjectWriter;
+
+import org.msgpack.jackson.dataformat.MessagePackFactory;
+
+/**
+ * Persists given state on the client with these properties:
+ * <ul>
+ * <li>State remains opaque (encrypted) so client cannot determine what is
+ * stored
+ * <li>State tampering is detected by using MAC
+ * <li>State timeout is enforced (default 15min)
+ * </ul>
+ * <p>
+ * Given a {@link KeySource} construct {@link ClientSideState} as:
+ * {@code new ClientSideState(keySource)}, and then persist state into HTTP
+ * Cookie with {@link #persist(String, String, Object)} method, and restore the
+ * state with {@link #restoreFrom(Cookie, Class)} method.
+ * <p>
+ * The implementation follows the
+ * <a href="https://tools.ietf.org/html/rfc6896">RFC6896</a> Secure Cookie
+ * Sessions for HTTP.
+ */
+public final class ClientSideState {
+    // 15 min
+    public static final long DEFAULT_TIMEOUT = 15 * 60;
+
+    private static final Decoder DECODER = Base64.getUrlDecoder();
+
+    private static final Encoder ENCODER = Base64.getUrlEncoder().withoutPadding();
+
+    private static final int IV_LEN = 16;
+
+    private static final ObjectMapper MAPPER = new ObjectMapper(new MessagePackFactory());
+
+    private final Edition edition;
+
+    private final BiFunction<Class<?>, byte[], Object> deserialization;
+
+    private final Supplier<byte[]> ivSource;
+
+    private final Function<Object, byte[]> serialization;
+
+    private final long timeout;
+
+    private final Supplier<Long> timeSource;
+
+    protected static final class RandomIvSource implements Supplier<byte[]> {
+        private static final ThreadLocal<SecureRandom> RANDOM = new ThreadLocal<SecureRandom>() {
```

Comment:
Yeah it is, this pattern is something I usually do, but now I remembered that we run in Linux containers so it will go to `/dev/urandom`, it would block otherwise on Windows and that makes it needed. Good catch 👍